### ISSUE TITLE
STklos crashes when trying to output characters into bytevector port with write

### ISSUE THREAD

**Issue by jpellegrini** [egallesio/STklos#159]
Hello @egallesio -- still while trying to port SRFI-207, I found another problem. So, R7RS says bytevector ports should work with `write-u8` and `read-u8`, and that works fine in STklos; but when `write` is used STklos crashes:

```
stklos> (call-with-port
 (open-output-bytevector)
 (lambda (out)
   (write #\a out)
   (write #\b out)
   (get-output-bytevector out)))
Received a SIGSGV signal.
Try to augment stack size (--stack-size option).
If the problem persists, send a mail to <bugs@stklos.net>
```

When `display` is used, then STklos raises the error and doesn't crash:

```
(call-with-port
 (open-output-bytevector)
 (lambda (out)
   (display #\b out)
   (display #\a out)
   (get-output-bytevector out)))
**** Error:
call-with-port: bad textual port `#[output-bytevector-port 7f43d64f6f20]'
```

One interesting thing: both Chibi and Gauche do accept `write` on bytevector ports... It would be an interesting addition to STklos in the future (automatically call `string->u8` and other auxiliary functions when writing to binary ports).

PS: something else is also strange when porting the reference implementation of SRFI-207; I'm looking into it.

**Comment by lassik** [comment]
`write` and `display` write a string; it's probably a good thing that they raise an error when trying to write to a binary port. Text and binary ports should be kept logically separate in most Scheme code. In this particular case, `open-output-string` would be the right thing to use instead of `open-output-bytevector`.

**Comment by lassik** [comment]
Strings have been a bit problematic in all languages since Unicode and other multibyte character encodings came along. Go strings are fundamentally bytevectors, and many standard library routines work on them assuming the bytes represent UTF-8. Scheme strings are fundamentally sequences of char objects; one char object is one codepoint. Some Schemes encode strings internally as UTF-16 (Kawa and IronScheme) or UTF-32 (Gambit).

**Comment by jpellegrini** [comment]
> `write` and `display` write a string; it's probably a good thing that they raise an error when trying to write to a binary port. Text and binary ports should be kept logically separate in most Scheme code. In this particular case, `open-output-string` would be the right thing to use instead of `open-output-bytevector`.

Ok good point. That's what I get for writing before having coffee. :)

Anyway, `write` crashes in STklos instead of raising an error -- PR #160 would fix this (STklos will then not crash, and raise an error when `write`ing to a binary port)

**Comment by lassik** [comment]
I agree that it shouldn't crash, and that raising an error is the right thing. Your patch looks like it solves the problem!

**Comment by lassik** [comment]
P.S. I don't think R7RS forbids `write` from working on binary ports. In general, the standard doesn't forbid things - it just stays silent and does not say anything about behavior that is non-standard. However, having non-standard behavior may cause surprises to people writing portable code, since the behavior won't work in other Scheme implementations, so it's good to err on the side of strictness.

**Comment by jpellegrini** [comment]
> Strings have been a bit problematic in all languages since Unicode and other multibyte character encodings came along. Go strings are fundamentally bytevectors, and many standard library routines work on them assuming the bytes represent UTF-8. Scheme strings are fundamentally sequences of char objects; one char object is one codepoint. Some Schemes encode strings internally as UTF-16 (Kawa and IronScheme) or UTF-32 (Gambit).


Right. Kawa and Gambit do also accept `write` on binary ports, and the result is different only for Kawa:

```
(call-with-port
 (open-output-bytevector)
 (lambda (out)
   (write #\a out)
   (write #\b out)
   (get-output-bytevector out)))
```

This gives `#u8(35 92 97 32 35 92 98)` in Kawa, and `#u8(35 92 97 35 92 98)` in Gauche, Chibi *and Gambit*!
I suppose that even representing strings as UTF-32, Gambit uses `string->utf8` or something similar to build the bytevector, and Kawa doesn't...

**Comment by lassik** [comment]
Wow, I had no idea it's so common to allow that. Thanks for investigating!

It does indeed look like Gambit is internally writing to a string and then using `string->utf8`.

**Comment by jpellegrini** [comment]
> P.S. I don't think R7RS forbids `write` from working on binary ports. In general, the standard doesn't forbid things - it just stays silent and does not say anything about behavior that is non-standard. However, having non-standard behavior may cause surprises to people writing portable code, since the behavior won't work in other Scheme implementations, so it's good to err on the side of strictness.

R7RS says that "_whether the textual and binary port types are disjoint is implementation-dependent_" (4th paragraph of 6.13.1), and I suppose this implies that textual procedures could work on binary ports (since a binary port could be textual anyway).

**Comment by lassik** [comment]
Good catch!

**Comment by jpellegrini** [comment]
> Your interpretation seems exactly right.

Which does not mean it's good to have the two kinds of port as the same. The PR I made actually works in the opposite direction (don't mix them!)

**Comment by egallesio** [comment]
Interesting discussion.

For me `write` is a function which permits to write a value that can be read back with the `read`.  The problem he e is that when you write in a bytevector, you don't have an easy way to get back the value (if  `write` has made such a conversion as the ones you showed above) . If  `(write #\a)`  produces `#u8(35 92 97)` there is no easy way to get back the `#\a` character (without converting the result to a string, converting it to a string port and reading in it).   

BTW, I have not the same interpretation of both of you for *R7RS says that "whether the textual and binary port types are disjoint is implementation-dependent" (4th paragraph of 6.13.1)* 
I read this as a port can be binary and textual at the same time. This is useful , for files which contains  Scheme values mixed with octets (a bytecode file for instance)
 In STklos, the standard ports are textual and binary to allow  `(write-bytevector #u8(65 66))` on stdin,as well as `(display "AB")`

In any case @jpellegrini , your fix in PR #160 is perfectly fine for me.
